numhask/src/NumHask/Analysis/Metric.hs

```diff
@@ -0,0 +1,366 @@
+--     | a > zero = one
+--     | otherwise = negate one
+--   abs = P.abs
+
+-- instance Signed Word where
+--   sign a
+--     | a == zero = zero
+--     | otherwise = one
+--   abs = P.abs
+
+-- instance Signed Word8 where
+--   sign a
+--     | a == zero = zero
+--     | otherwise = one
+--   abs = P.abs
+
+-- instance Signed Word16 where
+--   sign a
+--     | a == zero = zero
+--     | otherwise = one
+--   abs = P.abs
+
+-- instance Signed Word32 where
+--   sign a
+--     | a == zero = zero
+--     | otherwise = one
+--   abs = P.abs
+
+-- instance Signed Word64 where
+--   sign a
+--     | a == zero = zero
+--     | otherwise = one
+--   abs = P.abs
+
+-- | L1 and L2 norms are provided for potential speedups, as well as the generalized p-norm.
+--
+-- for p >= 1
+--
+-- > normLp p a >= zero
+-- > normLp p zero == zero
+--
+-- Note that the Normed codomain can be different to the domain.
+--
+class Normed a b where
+  normL1 :: a -> b
+  normL2 :: a -> b
+  normLp :: b -> a -> b
+
+instance Normed Double Double where
+  normL1 = P.abs
+  normL2 = P.abs
+  normLp _ a = P.abs a
+
+instance Normed Float Float where
+  normL1 = P.abs
+  normL2 = P.abs
+  normLp _ a = P.abs a
+
+instance Normed Int Int where
+  normL1 = P.abs
+  normL2 = P.abs
+  normLp _ a = P.abs a
+
+instance Normed Integer Integer where
+  normL1 = P.abs
+  normL2 = P.abs
+  normLp _ a = P.abs a
+
+instance (Multiplication a, ExpField a, Normed a a) =>
+         Normed (Complex a) a where
+  normL1 (rx :+ ix) = normL1 rx + normL1 ix
+  normL2 (rx :+ ix) = sqrt (rx * rx + ix * ix)
+  normLp p (rx :+ ix) = (normL1 rx ** p + normL1 ix ** p) ** (one / p)
+
+instance Normed Natural Natural where
+  normL1 = P.abs
+  normL2 = P.abs
+  normLp _ a = P.abs a
+
+instance Normed Int8 Int8 where
+  normL1 = P.abs
+  normL2 = P.abs
+  normLp _ a = P.abs a
+
+instance Normed Int16 Int16 where
+  normL1 = P.abs
+  normL2 = P.abs
+  normLp _ a = P.abs a
+
+instance Normed Int32 Int32 where
+  normL1 = P.abs
+  normL2 = P.abs
+  normLp _ a = P.abs a
+
+instance Normed Int64 Int64 where
+  normL1 = P.abs
+  normL2 = P.abs
+  normLp _ a = P.abs a
+
+instance Normed Word Word where
+  normL1 = P.abs
+  normL2 = P.abs
+  normLp _ a = P.abs a
+
+instance Normed Word8 Word8 where
+  normL1 = P.abs
+  normL2 = P.abs
+  normLp _ a = P.abs a
+
+instance Normed Word16 Word16 where
+  normL1 = P.abs
+  normL2 = P.abs
+  normLp _ a = P.abs a
+
+instance Normed Word32 Word32 where
+  normL1 = P.abs
+  normL2 = P.abs
+  normLp _ a = P.abs a
+
+instance Normed Word64 Word64 where
+  normL1 = P.abs
+  normL2 = P.abs
+  normLp _ a = P.abs a
+
+-- | distance between numbers using L1, L2 or Lp-norms
+--
+-- > distanceL2 a b >= zero
+-- > distanceL2 a a == zero
+-- > \a b c -> distanceL2 a c + distanceL2 b c - distanceL2 a b >= zero &&
+-- >           distanceL2 a b + distanceL2 b c - distanceL2 a c >= zero &&
+-- >           distanceL2 a b + distanceL2 a c - distanceL2 b c >= zero &&
+class Metric a b where
```

Comment:
I'm not sure about this class. Not all metric spaces are induced by a Lebesgue norm.

Reply:
A "metric" is usually taken to be a binary function that's:
* non-negative
* commutative
* sub-additive

Reply:
I agree.

However, this definition of Metric does not appear to be introduced by @LeanderK but just appears in the diff because things have been moved around. In the interests of keeping the evaluation of this PR as narrow as possible, I've opened a separate issue (#81) to discuss Metrics.

Reply:
I am currently writing a very "rigorous" linear algebra library on top of numhask and there I use a proper metric-definition. I would tend to leave the Metric as it is, because the alternative is way more complex to define.

Reply:
I redefine some numhask definitions, but they tend to be way more complex.